### PR TITLE
[Refactor] Adding particles properties

### PR DESCRIPTION
**Pull request by bodhinandach**
**Describe the PR**
Adding container and general functions in node and particle_base for fluid and two-phase particle purposes.

**Related Issues/PRs**
#637, #650, and #654

**Additional context**
Due to the observed performance issue, the current containers and general functions in `node` and `particle_base` are added for `fluid` and `two_phase` particles without changing the solid_particle. Test should be added later in the PR of fluid and two-phase particles.

### CODE REVIEW

File: include/node.h [hunk 1]
```diff
@@ -65,6 +65,35 @@ class Node : public NodeBase<Tdim> {
...
+  //! \param[in] update A boolean to update (true) or assign (false)
+  //! \param[in] phase Index corresponding to the phase
+  //! \param[in] value Property value from the particles in a cell
+  void update_scalar_property(mpm::properties::Scalar property, bool update,
+                              unsigned phase, double value) noexcept override;
+
+  //! Return property at a given node for a given phase
+  //! \param[in] property Name of the property to return
+  //! \param[in] phase Index corresponding to the phase
+  double scalar_property(mpm::properties::Scalar property,
```
Comment (kks32) [inline include/node.h:92]: We need a retval

File: include/node.h [hunk 2]
```diff
@@ -65,6 +65,35 @@ class Node : public NodeBase<Tdim> {
...
+  //! \param[in] phase Index corresponding to the phase
+  //! \param[in] value Property value from the particles in a cell
+  virtual void update_vector_property(
+      mpm::properties::Vector property, bool update, unsigned phase,
+      const Eigen::Matrix<double, Tdim, 1>& value) noexcept override;
+
+  //! Return property at a given node for a given phase
+  //! \param[in] property Name of the property to return
+  //! \param[in] phase Index corresponding to the phase
+  virtual Eigen::Matrix<double, Tdim, 1> vector_property(
```
Comment (kks32) [inline include/node.h:108]: Add a \retval

File: include/node_base.h [hunk 3]
```diff
@@ -70,6 +72,36 @@ class NodeBase {
...
+  //! \param[in] phase Index corresponding to the phase
+  //! \param[in] value Property value from the particles in a cell
+  virtual void update_scalar_property(mpm::properties::Scalar property,
+                                      bool update, unsigned phase,
+                                      double value) noexcept = 0;
+
+  //! Return property at a given node for a given phase
+  //! \param[in] property Name of the property to return
+  //! \param[in] phase Index corresponding to the phase
+  virtual double scalar_property(mpm::properties::Scalar property,
```
Comment (kks32) [inline include/node_base.h:99]: \retval

File: include/particles/particle_base.h [hunk 4]
```diff
@@ -124,6 +124,73 @@ class ParticleBase {
...
+  //! Update scalar property at the particle
+  //! \param[in] property Name of the property to update
+  //! \param[in] update A boolean to update (true) or assign (false)
+  //! \param[in] value Property value from the particles in a cell
+  void update_scalar_property(mpm::properties::Scalar property, bool update,
+                              double value) noexcept;
+
+  //! Return property
+  //! \param[in] phase Index corresponding to the phase
+  double scalar_property(mpm::properties::Scalar property) const;
```
Comment (kks32) [inline include/particles/particle_base.h:151]: Add \retval

File: include/particles/particle_base.h [hunk 5]
```diff
@@ -124,6 +124,73 @@ class ParticleBase {
...
+  //! \param[in] update A boolean to update (true) or assign (false)
+  //! \param[in] phase Index corresponding to the phase
+  //! \param[in] value Scalar value to be mapped from particle to node
+  void map_scalar_property_nodes(mpm::properties::Scalar property, bool update,
+                                 unsigned phase, double value) noexcept;
+
+  //! Return an interpolation of scalar property in particle from nodes
+  //! \param[in] property Name of the property to update
+  //! \param[in] phase Index corresponding to the phase
+  double interpolate_scalar_property_nodes(mpm::properties::Scalar property,
```
Comment (kks32) [inline include/particles/particle_base.h]: Add \retval

File: include/particles/particle_base.tcc [hunk 6]
```diff
@@ -15,3 +15,106 @@ mpm::ParticleBase<Tdim>::ParticleBase(Index id, const VectorDim& coord,
...
+
+  // Map scalar property to nodes
+  for (unsigned i = 0; i < nodes_.size(); ++i)
+    nodes_[i]->update_scalar_property(
+        property, update, phase, scalar_properties_.at(property) * shapefn_[i]);
+}
+
+//! Map an arbitrary scalar value to nodal scalar property
+template <unsigned Tdim>
+void mpm::ParticleBase<Tdim>::map_scalar_property_nodes(
```
Comment (kks32) [inline include/particles/particle_base.tcc]: could we add `from` or `to` to the function names to make it clear?

**Review by kks32 — CHANGES_REQUESTED** [review]
Thanks for the PR,  could you please make the minor cosmetic changes.

**Review by ezrayst — COMMENTED** [review]
Thanks for this. I believe this is option (3) right from #654?

I have one question on this implementation in regards to `hdf5` and `resume` function. I guess there are enums and the IO would be done with no problem right?

**Comment by bodhinandach** [comment]
@ezrayst This is for option (2). The option (3) is the PR #654 itself. As for hdf5 and resume, since the solid particle is untouched here, there will be no problem.

**Comment by ezrayst** [comment]
Alright, but what about the `hdf5` and `resume` for fluid phase for instance?

**Comment by bodhinandach** [comment]
> Alright, but what about the `hdf5` and `resume` for fluid phase for instance?

This will be implemented in the fluid or two-phase particles in the future PRs.

**Comment by bodhinandach** [comment]
@kks32 Should we delete this branch and archive `refactor/particles`?

**Pull request closed without merging**